[MRG+1] Add example for GroupShuffleSplit

JesperDramsch opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
OK
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Toward #3846 
at Euroscipy Sprint with @adrinjalali 

this PR adds a long example to `model_selection.GroupShuffleSplit

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

JesperDramsch committed:
Add example for GroupShuffleSplit

JesperDramsch committed:
Removed Lint

adrinjalali commented on sklearn/model_selection/_split.py:
```diff
@@ -1481,6 +1481,50 @@ class GroupShuffleSplit(ShuffleSplit):
+    >>> X = np.array([[1, 2], [3, 4], [5, 6]])
+    >>> y = np.array([1, 2, 1])
+    >>> groups = np.array([1, 2, 3, 4, 5])
+    >>> gss = GroupShuffleSplit()
```
you need to always provide `random_state` where it's a parameter, to be sure it's deterministic.

adrinjalali commented on sklearn/model_selection/_split.py:
```diff
@@ -1481,6 +1481,50 @@ class GroupShuffleSplit(ShuffleSplit):
+    TRAIN: [1 2] TEST: [0]
+    [[3 4]
+     [5 6]] [[1 2]] [2 1] [1]
+    >>> gss = GroupShuffleSplit(n_splits=2, train_size=.5, random_state=42)
```
I'm not sure, but I think only one of the examples.

adrinjalali reviewed (commented):
Otherwise looks good to me, thanks @JesperDramsch

JesperDramsch committed:
Reduce Examples

JesperDramsch committed:
Fix Formatting Error

JesperDramsch committed:
Fixed Groups in Example

adrinjalali reviewed (approved):
LGTM, thanks @JesperDramsch

thomasjpfan commented on sklearn/model_selection/_split.py:
```diff
@@ -1481,6 +1481,27 @@ class GroupShuffleSplit(ShuffleSplit):
+    >>> from sklearn.model_selection import GroupShuffleSplit
+    >>> X = np.array([[1, 2], [3, 4], [5, 6]])
+    >>> y = np.array([1, 2, 1])
+    >>> groups = np.array([1, 2, 3])
```
Should we showcase adjust this example to showcase what happens when there is more than one sample in a group? For example:

```py
import numpy as np
from sklearn.model_selection import GroupShuffleSplit
X = np.ones(shape=(8, 2))
y = np.array([0, 1, 0, 1, 0, 1, 0, 1])
groups = np.array([1, 1, 1, 2, 2, 3, 3, 3])
gss = GroupShuffleSplit(n_splits=2, train_size=.6, random_state=42)
for train_idx, test_idx in gss.split(X, y, groups):
    print("TRAIN:", train_idx, "TEST:", test_idx)
```

rth replied:
+1 also showing `train_idx`, and `test_idx` is enough I think, the actual content of `X`, `y` doesn't matter for this example.

JesperDramsch replied:
I honestly just had a look at how other model selection examples are presented. 

`LeavePGroupsOut` was my main source in this case.

JesperDramsch replied:
@thomasjpfan / @rth Gotta say this is a more insightful example. Should I check if I can adjust `LeavePGroupsOut`?

JesperDramsch committed:
Reduce Example to essentials

Implement suggested changes by Core Devs to reduce examples to essentials consistent with ShuffleSplit

thomasjpfan commented:
Thank you @JesperDramsch !
